support integer arrays comparison

xh4n3 opened the pull request:
This PR implements the binop comparison for integer arrays, it is a part of #2248 , a follow-up to #2261 .

By using this binop and the `pton` built-in together, we can now select packets filtered by addresses. Example:

```
tracepoint:tcp:tcp_retransmit_skb {
  $addr = pton("::1");
  if (args->daddr_v6 == $addr) {
    printf("localhost ipv6 retransmit\n");
  } else {
    printf("non-localhost ipv6 retransmit\n");
  }
  return;
}
```

Comments are welcome.

##### Checklist
- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1336,6 +1336,38 @@ void CodegenLLVM::binop_string(Binop &binop)
+  // cmp returns 0 when arrays are equal
+  bool inverse = binop.op == Operator::EQ;
+
+  auto left_array_ptr = getString(binop.left);
+  auto right_array_ptr = getString(binop.right);
```
This doesn't seem correct, if you want to work with integer arrays the why use `getString`?

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1336,6 +1336,38 @@ void CodegenLLVM::binop_string(Binop &binop)
+               << "\"";
+  }
+
+  std::string string_literal;
```
Unused variable?

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,71 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+    offset = CreateMul(getInt64(i), getInt64(elem_size));
+    l_addr = CreateAdd(val1, offset);
+    r_addr = CreateAdd(val2, offset);
+    CreateProbeRead(ctx, l, getInt32(1), l_addr, AddrSpace::kernel, loc);
```
If the value is on stack already (e.g. when using `pton`) then you should use GEP instead of proberead.

xh4n3 replied:
@viktormalik Thanks! So the proper way is, use a GEP when the value is on stack, use a proberead when it is not. Am I understanding it correctly? How should determine if the value is on stack, use IsCtxAccess?

viktormalik replied:
Sorry for a late answer. There's a function `shouldBeOnStackAlready` which should be used to detect this (although you'll have to add `pton` there for your usecase).

xh4n3 replied:
@viktormalik Cool, will try that. Besides, may i know what does the CtxAccess mean?

viktormalik replied:
Context is a special structure that is passed to each BPF program. The exact information passed within the context depends on the program type.

xh4n3 replied:
Thanks. I found `shouldBeOnStackAlready` only recognizes those specific `SizedType`, while `pton` is returning byte arrays only. May I use `is_internal` instead?

viktormalik replied:
Yeah, `is_internal` makes sense. And you can use `onStack` instead of `shouldBeOnStackAlready`, then.

xh4n3 replied:
@viktormalik Great, now the code is using `is_internal`. Please let me if any other issues. Thanks!

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,71 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+
+  Function *parent = GetInsertBlock()->getParent();
+  AllocaInst *store = CreateAllocaBPF(getInt1Ty(), "arraycmp.result");
+  BasicBlock *str_ne = BasicBlock::Create(module_.getContext(),
```
`arr_ne` would be a better name here

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,71 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+  CreateStore(getInt1(!inverse), store);
+  for (size_t i = 0; i < n; i++)
+  {
+    BasicBlock *char_eq = BasicBlock::Create(module_.getContext(),
```
Same as above: `arr_eq` is a more suitable name.

viktormalik commented on tests/runtime/array:
```diff
@@ -142,3 +142,15 @@ RUN {{BPFTRACE}} -e 'struct A { int x[4]; } uprobe:./testprogs/array_access:test
 AFTER ./testprogs/array_access
+
+NAME array compare eq
+PROG struct A { int x[4]; } uprobe:./testprogs/array_access:test_arrays { $a = (struct A *) arg0; $b = (struct A *) arg0; if ($a == $b) { printf("two int arrays are equal.\n"); } exit(); }
```
This is not correct, you're comparing pointers (`struct A *`) instead of arrays. Should be `$a->x == $b->x`.

viktormalik reviewed (commented):
It feels like we're missing some checks in the semantic analyser when comparing arrays - in particular equality of inner types and of the array sizes.

xh4n3 commented:
@viktormalik Thank you for reviewing this. 😃 

> It feels like we're missing some checks in the semantic analyser when comparing arrays - in particular equality of inner types and of the array sizes.

I've added this part.

> If the value is on stack already (e.g. when using pton) then you should use GEP instead of proberead.

I've updated this part also, not sure if I understand it correctly. Could you please help me check it?
https://github.com/iovisor/bpftrace/pull/2387/files#diff-3d80d60dba9cfc34c2fe79c01df26172886685f76475c9b53b372136e49f1af0R876-R915

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,103 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+    BasicBlock *arr_eq = BasicBlock::Create(module_.getContext(),
+                                            "arraycmp.loop",
+                                            parent);
+    if (onStack(array_type))
```
This doesn't seem correct. One of the arrays may be on stack while the other one is not.

I'm also wondering if it'd make sense to make use of the `readDatastructElemFromStack` and `probereadDatastructElem` functions.

xh4n3 replied:
> if it'd make sense to make use of the readDatastructElemFromStack and probereadDatastructElem functions.

This is doable, then we should be moving `readDatastructElemFromStack`  and `probereadDatastructElem` to `irbuilderbpf.cpp`. But the IRBuilderBPF doesn't have the `current_attach_point` variable, so I guess I need to pass this as a parameter. Is that alright?

viktormalik replied:
Yeah, that would be necessary and also, `expr_deleter_` is missing. So I guess that it's best to just not use those functions (they anyway do more functionality than what is necessary here).

viktormalik commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -1675,6 +1676,29 @@ void SemanticAnalyser::binop_int(Binop &binop)
+        << "Only arrays of integer support comparison operators.";
+  }
+
+  if (lht.GetSize() != rht.GetSize())
```
This should compare the number of elements and the element size separately.

viktormalik reviewed (changes requested):
This feature should also have a codegen test.

viktormalik commented on tests/codegen/array_integer_not_equal_comparison.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "common.h"
```
No need to have two (almost) identical codegen tests.

xh4n3 commented:
> This feature should also have a codegen test.

Done.

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,101 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+          ctx, l, getInt32(elem_type.GetSize()), ptr_l, val1_type.GetAS(), loc);
+      ll = CreateLoad(GetType(elem_type), l);
+    }
+    if (onStack(val2_type))
+    {
+      val2 = CreateIntToPtr(val2, GetType(val1_type)->getPointerTo());
+      auto *ptr_r = CreateGEP(GetType(val1_type),
+                              val2,
+                              { getInt32(0), getInt32(i) });
+      rr = CreateLoad(GetType(elem_type), ptr_r);
+    }
+    else
+    {
+      auto *ptr_r = CreateAdd(
+          val2, CreateMul(getInt64(i), getInt64(elem_type.GetSize())));
+      CreateProbeRead(
+          ctx, r, getInt32(elem_type.GetSize()), ptr_r, val1_type.GetAS(), loc);
+      rr = CreateLoad(GetType(elem_type), r);
+    }
```
TBH, I'm not very keen that this is basically a repetition of the above code. As we discussed, this is a special use-case which doesn't need full-fledged element accessing functions, so I guess I'm ok with this. But still, if you found a way to factor this into a reasonable function, that would be great.

viktormalik replied:
It does look better, but now there's a confusion whether to use `CreateGetElemInArrayOnStack` or `readDatastructElemFromStack`. They do similar things, although the new one is specialized for non-context integer-only accesses (same applies for proberead variants).

It's probably the best to leave the original expanded code here - it's a single function after all. Or, alternatively, you could move it to `codegen_llvm.cpp` in which case you'll be able to use the current element accessing functions.

xh4n3 replied:
> Or, alternatively, you could move it to codegen_llvm.cpp in which case you'll be able to use the current element accessing functions.

Do you mean that I could move the `CreateGetElemInArrayOnStack` function into `codegen_llvm.cpp`?
Then `IRBuilderBPF`'s `CreateIntegerArrayCmp` would be calling `CodegenLLVM`'s methods, which I don't think it is a good idea.

xh4n3 replied:
I see, that would leave `CreateIntegerArrayCmp` and `CreateStrncmp` in separate files, kind of strange. Let me keep the expanded one for this PR.

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1336,6 +1336,42 @@ void CodegenLLVM::binop_string(Binop &binop)
+  // integer array compare returns 0 when arrays are equal
+  bool inverse = binop.op == Operator::EQ;
+
+  auto scoped_del_right = accept(binop.left);
```
```suggestion
  auto scoped_del_left = accept(binop.left);
```

xh4n3 replied:
Sorry, I was being stupid..

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1336,6 +1336,42 @@ void CodegenLLVM::binop_string(Binop &binop)
+  auto scoped_del_right = accept(binop.left);
+  Value *left_array_val = expr_;
+
+  auto scoped_del_left = accept(binop.right);
```
```suggestion
  auto scoped_del_right = accept(binop.right);
```

viktormalik commented:
This has quite a lot of code, I'd like to have one more review.

@danobi, @ajor, @fbs, @BurntBrunch

danobi commented:
Missing semantic analysis unit tests? Maybe also test semantic analysis errors out correctly for:
* arrays of different length
* arrays of different type
* array compared to !array

BurntBrunch commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,101 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+  CreateStore(getInt1(!inverse), store);
+  for (size_t i = 0; i < n; i++)
+  {
+    BasicBlock *arr_eq = BasicBlock::Create(module_.getContext(),
```
Similarly to the strstr discussion in #2393, I think this implementation would be better if we don't unroll the loop.

danobi replied:
Seems reasonable to me to gate this helper based on bounded loop presence.

BurntBrunch replied:
> based on bounded loop support

That would be.. [5.3, the patch is from Jun 2019](https://lwn.net/Articles/794934/). Runtime checks for this are annoying but possible.

danobi replied:
You mean a feature probe like this? https://github.com/iovisor/bpftrace/blob/86b7843b0e56c96dc91c9fa0354ffe1507245ad8/src/bpffeature.cpp#L191-L207

xh4n3 commented:
Hi @danobi , thanks for the review!
I've added semantic analysis tests and an unrolling-disabled variation for the `CreateIntegerArrayCmp` method.

viktormalik commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,238 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+  SetInsertPoint(while_cond);
+  auto *cond = CreateICmpSLT(CreateLoad(getInt32Ty(), i),
+                             CreateLoad(getInt32Ty(), n),
+                             "true_cond");
```
Maybe
```suggestion
                             "size_check");
```
`true_cond` doesn't seem like a good name.

viktormalik reviewed (commented):
Thanks for the update! The loop variant looks good, although I've couple of nits to improve the IR. Feel free to ignore those that don't make sense to you.

xh4n3 commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,238 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+  return result;
+}
+
+Value *IRBuilderBPF::CreateIntegerArrayCmp(Value *ctx,
```
I renamed the unrolled function to `CreateIntegerArrayCmpUnrolled`. I think if this is the right way to go, we probably should refactor other irbuilder functions as well, e.g. the `strncmp`.

xh4n3 commented on src/ast/irbuilderbpf.cpp:
```diff
@@ -830,6 +830,238 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
+  CreateBr(while_cond);
+  SetInsertPoint(while_end);
+
+  CreateBr(done);
```
I have to confess that I borrowed some lines of codes from other irbuilder functions, including those empty line code style.
Thanks for letting me understand this. Let me create a PR to clean up these bad code style.

viktormalik replied:
I'm not a big fan of code-style-only changes (makes blaming harder). Let's leave the rest as is and change it once we're doing functional changes (which we'll probably do for some, e.g. `strncmp`).

xh4n3 replied:
Okay, I see.

fbs commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -1682,6 +1683,35 @@ void SemanticAnalyser::binop_int(Binop &binop)
+        << "The " << opstr(binop) << " operator cannot be used on arrays.";
+  }
+
+  if (!lht.GetElementTy()->IsIntegerTy() || !rht.GetElementTy()->IsIntegerTy())
```
maybe we can just do:

```
  if (lht.GetNumElements() != rht.GetNumElements())
  {
    LOG(ERROR, binop.loc, err_)
     << "Arrays must be of the same length. " << lth << " != " << rht;
  }

if (!lht.GetElementTy()->IsIntegerTy()  || lht != rht) { // uses the IsEqual method on sized type
  LOG(ERROR, binop.loc, err_)
     << "Array elements must be of the same integer type. " << lth << " != " << rht;
}
```

This removes 2 related checks and also indicates what the mismatching types are

xh4n3 replied:
Sure!

fbs reviewed (changes requested):
looks good!

xh4n3 commented:
@fbs @danobi Hi, I've done the changes regarding to your comments. Could I have an approval? Thanks!

fbs commented:
great work! I'll merge it soon

fbs commented:
woops the other merge caused a conflict :(

xh4n3 commented:
> woops the other merge caused a conflict :(

fixed :p

fbs commented:
thanks!
